File: python/ex2/raster.py
import numpy as np

class Raster:
    def __init__(self, w, h, data, gap_count=-1):
        if w <= 0 or h <= 0 or data.size != w * h:
            raise ValueError();
        self.w = w
        self.h = h
        self.size = w * h
        self.data = data
        if gap_count >= 0:
            self.gap_count = gap_count;
        else:
            c = 0
            for i in range(data.size):
                if np.isnan(data[i]):
                    c += 1
            self.gap_count = c

    def is_singular(self):
        return self.w == 1 and self.h == 1

    def is_free_of_gaps(self):
        return self.gap_count == 0

    def is_full_of_gaps(self):
        return self.gap_count == self.size

    def clone(self):
        return Raster(self.w, self.h, np.copy(self.data), self.gap_count)

File: java/src/main/java/com/bc/perfcomp/ex2/Raster.java
package com.bc.perfcomp.ex2;

/**
 * @author Norman Fomferra
 */
final class Raster implements Cloneable {
    final int w, h;
    final double[] data;
    final int gapCount;

    public Raster(int w, int h, double[] data) {
        this(w, h, data, countGaps(data));
    }

    public Raster(int w, int h, double[] data, int gapCount) {
        this.w = w;
        this.h = h;
        this.data = data;
        this.gapCount = gapCount;
    }

    public boolean isFreeOfGaps() {
        return gapCount == 0;
    }

    public boolean isFullOfGaps() {
        return gapCount == w * h;
    }

    public boolean isSingular() {
        return w == 1 && h == 1;
    }

    public static boolean isGap(double v) {
        return Double.isNaN(v);
    }

    public static int countGaps(double[] data) {
        int gapCount = 0;
        for (double v : data) {
            if (isGap(v)) {
                gapCount++;
            }
        }
        return gapCount;
    }

    @Override
    public Raster clone() {
        return new Raster(w, h, data.clone(), gapCount);
    }

}
